### PR TITLE
Dialog floated for forms.

### PR DESCRIPTION
**Pull request by nobuti**
This PR implements #7786.

modals

It adds the option to render dialog controls floated, so it's shown above all the thing in the UI. Even without being in the same context that the field, it should keep the position if the user does scroll. If there is no space to follow, it will get anchored at the top or bottom of the window.

### How to test
- [ ] Check that all dialog controls are working properly.
- [ ] Check if having the option `dialogMode: 'float'`, it's rendered in the global context, this is, the body.
- [ ] Double check all the analysis, style, popup, legend, and feature form are working properly.
- [ ] Check that once the form is removed, no orphan DOM node remains in the HTML.
- [ ] Check all dialog controls have incorporated this feature.

### CODE REVIEW

File: lib/assets/test/spec/SpecHelper3.js [hunk 1]
```diff
@@ -152,6 +161,12 @@ afterEach(function () {
     this.view.clean();
   }
 
+  // Delete all dropdowns and run
```
Comment (xavijam) [inline lib/assets/test/spec/SpecHelper3.js]: Not needed then?

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill.js [hunk 2]
```diff
@@ -131,7 +136,10 @@ Backbone.Form.editors.Fill = Backbone.Form.editors.Base.extend({
     this._fillDialogView.render();
     this._fillDialogView.show();
 
-    this.$el.append(this._fillDialogView.$el);
+    // this.$el.append(this._fillDialogView.$el);
```
Comment (matallo) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill.js]: 🔥

File: lib/assets/core/javascripts/cartodb3/components/popup-manager.js [hunk 3]
```diff
@@ -0,0 +1,134 @@
+var _ = require('underscore');
```
Comment (matallo) [inline lib/assets/core/javascripts/cartodb3/components/popup-manager.js]: is "popup-manager" the best naming for this? it makes me think of the CARTO popups

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/fill-dialog.js [hunk 4]
```diff
@@ -3,7 +3,7 @@ var CoreView = require('backbone/core-view');
 module.exports = CoreView.extend({
   className: 'Editor-boxModal Editor-FormDialog js-formDialog is-opening',
 
-  initialize: function () {
+  initialize: function (opts) {
```
Comment (matallo) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/fill-dialog.js]: why do we need the `opts` if we are not making use of them?

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/operators/operators-view.js [hunk 5]
```diff
@@ -48,7 +50,12 @@ Backbone.Form.editors.Operators = Backbone.Form.editors.Base.extend({
       collection: this.collection
     });
     this._operatorsListView.bind('change', this._onOperatorsChange, this);
-    this.$el.append(this._operatorsListView.el); // No render from the beginning
+
+    // this.$el.append(this._operatorsListView.el); // No render from the beginning
```
Comment (matallo) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/operators/operators-view.js]: 🔥

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/operators/operators-view.js [hunk 6]
```diff
@@ -68,9 +75,11 @@ Backbone.Form.editors.Operators = Backbone.Form.editors.Base.extend({
   _initBinds: function () {
     this.applyESCBind(function () {
       this._operatorsListView.hide();
```
Comment (matallo) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/operators/operators-view.js]: can't we make a method for these two, which we are repeating some lines below?

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/select/multi-select-view.js [hunk 7]
```diff
@@ -88,16 +91,21 @@ Backbone.Form.editors.MultiSelect = Backbone.Form.editors.Base.extend({
 
     this._listView.bind('hidden', this._onHide, this);
 
-    this.$el.append(this._listView.el); // No render from the beginning
+    // this.$el.append(this._listView.el); // No render from the beginning
```
Comment (matallo) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/select/multi-select-view.js]: 🔥

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/select/multi-select-view.js [hunk 8]
```diff
@@ -88,16 +91,21 @@ Backbone.Form.editors.MultiSelect = Backbone.Form.editors.Base.extend({
...
+    // document.body.appendChild(this._listView.el);
+    this._popupManager = new PopupManager(this.cid, this.$el, this._listView.$el);
+    this._popupManager.append(this.dialogMode);
   },
 
   _initBinds: function () {
     this.collection.bind('change:selected', this._onItemSelected, this);
     this.applyESCBind(function () {
       this._listView.hide();
```
Comment (matallo) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/select/multi-select-view.js]: same than ^^^

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/select/select-view.js [hunk 9]
```diff
@@ -76,27 +79,34 @@ Backbone.Form.editors.Select = Backbone.Form.editors.Base.extend({
       position: this.options.position
     });
 
-    this.$el.append(this._listView.el); // No render from the beginning
+    // this.$el.append(this._listView.el);
```
Comment (matallo) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/select/select-view.js]: 🔥

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/select/select-view.js [hunk 10]
```diff
@@ -76,27 +79,34 @@ Backbone.Form.editors.Select = Backbone.Form.editors.Base.extend({
...
+    // document.body.appendChild(this._listView.el);
+    this._popupManager = new PopupManager(this.cid, this.$el, this._listView.$el);
+    this._popupManager.append(this.dialogMode);
   },
 
   _initBinds: function () {
     this.collection.bind('change:selected', this._onItemSelected, this);
     this.applyESCBind(function () {
       this._listView.hide();
```
Comment (matallo) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/select/select-view.js]: same than ^^^

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/select/suggest-view.js [hunk 11]
```diff
@@ -92,28 +95,35 @@ Backbone.Form.editors.Suggest = Backbone.Form.editors.Base.extend({
       position: this.options.position
     });
 
-    this.$el.append(this._listView.el); // No render from the beginning
+    // this.$el.append(this._listView.el); // No render from the beginning
```
Comment (matallo) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/select/suggest-view.js]: 🔥

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/select/suggest-view.js [hunk 12]
```diff
@@ -92,28 +95,35 @@ Backbone.Form.editors.Suggest = Backbone.Form.editors.Base.extend({
...
+    this._popupManager = new PopupManager(this.cid, this.$el, this._listView.$el);
+    this._popupManager.append(this.dialogMode);
   },
 
   _initBinds: function () {
     this.collection.bind('change:selected', this._onItemSelected, this);
     this.collection.bind('reset', this.render, this);
     this.applyESCBind(function () {
       this._listView.hide();
+      this._popupManager.untrack();
```
Comment (matallo) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/select/suggest-view.js]: same than ^^^

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/select/select-view.js [hunk 13]
```diff
@@ -138,6 +150,7 @@ Backbone.Form.editors.Select = Backbone.Form.editors.Base.extend({
   },
 
   remove: function () {
+    this._popupManager.destroy();
```
Comment (matallo) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/select/select-view.js]: in other components we're checking `this._popupManager && this._popupManager.destroy();` why not here?

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/select/multi-select-view.js [hunk 14]
```diff
@@ -160,6 +172,7 @@ Backbone.Form.editors.MultiSelect = Backbone.Form.editors.Base.extend({
   },
 
   remove: function () {
+    this._popupManager.destroy();
```
Comment (matallo) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/select/multi-select-view.js]: in other components we're checking `this._popupManager && this._popupManager.destroy();` why not here?

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/operators/operators-view.js [hunk 15]
```diff
@@ -116,6 +129,7 @@ Backbone.Form.editors.Operators = Backbone.Form.editors.Base.extend({
   },
 
   remove: function () {
+    this._popupManager.destroy();
```
Comment (matallo) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/operators/operators-view.js]: in other components we're checking `this._popupManager && this._popupManager.destroy();` why not here?

File: lib/assets/core/javascripts/cartodb3/components/popup-manager.js (hunk 3 above)
Comment (nobuti) [inline lib/assets/core/javascripts/cartodb3/components/popup-manager.js]: Feel free to suggest :)

**Comment by matallo** [comment]
this is already available in ded04 staging

**Comment by javierarce** [comment]
### Acceptance

In general it works wonderfully, but I've found an issue when the fill component changes of size (for example, when navigating from the color picker view to a list of categories with a couple of elements).

popup

As you can see, the dialog changes the position very abruptly and is annoying :(

**Comment by nobuti** [comment]
Agreed, we can block the position on size changes. I'll fix it asap.

**Comment by nobuti** [comment]
@javierarce we try to position it in the same direction as it was unless there is no space. In that case, I guess it's better to change the position. Let me know if you don't think so. I've just pushed the fix for the inputs for the fill component too.

**Comment by matallo** [comment]
👍 still leaving it in CR in case @javierarce wants to make any comment, but feel free to move to acceptance

**Comment by javierarce** [comment]
Thanks! Ready to go to Acceptance.

**Comment by xavijam** [comment]
💃 , hey @ivanmalagon, could you share your testing user with @noguerol 💋  ?

**Comment by ivanmalagon** [comment]
Acceptance done in staging.

It works like a charm (kudos to @nobuti , great job in this tricky issue). Nevertheless, I found two problems.

1. Not working in Custom legend color picker.

image

2. Adding `predict trends` or `filter by layer` analyses throws an error related to this PR. Digging in the stack trace, the error takes place when removing a dialog. It's likely that this happens because these two analyses share a form element that breaks.

image

3. Date picker in feature edition not on top.

image

Not tested:
- DOM check.
- `dialogMode: 'float'`

I'll test this now but I wanted to point those found issues.

**Comment by nobuti** [comment]
I think cases 1 and 3 are because they are `nested`. I'll take a look asap.

**Comment by ivanmalagon** [comment]
Another issue found.

After opening lots of dialogs in `fill` and `labels`, I switch to another aggregation options. Then I get console errors trown by `popup-manager`

**Comment by nobuti** [comment]
Ready for another round.

**Comment by ivanmalagon** [comment]
Rock solid. Waiting for all the checks to pass again before merging.

ok

**Comment by nobuti** [comment]
\o/

**Comment by xavijam** [comment]
A before and an after in the history of CARTO Builder.

**Comment by jorgesancha** [comment]
Top job and beautiful team work :1st_place_medal:

**Pull request merged**